crates/dbs-miniball/README.md

```diff
@@ -1,21 +1,528 @@
+```
+
+4. Configure the vCPUs. This is done through `vm-vcpu` crate, which for now is a
+   local crate, but once the interfaces there get stabilized, they will be
+   upstreamed in [`vmm-vcpu`](https://github.com/rust-vmm/vmm-vcpu).
+    
+```rust
+// src/vmm/src/lib.rs
+
+// Create the KvmVm.
+let vm_config = VmConfig::new(&kvm, config.vcpu_config.num)?;
+```
+
+   1. Requirements: KVM is configured, guest memory is configured
+   2. Inputs: vCPU registry values - hardcoded / embedded in VMM for the same
+      reasons as boot parameters.
+   3. Breakdown (`x86_64`):
+       1. Configure MPTables. These
+        [tables](https://pdos.csail.mit.edu/6.828/2014/readings/ia32/MPspec.pdf)
+        tell the guest OS what the multiprocessor configuration looks like,
+        and are required even with a single vCPU.
+
+        ```rust
+        // src/vm-vcpu/src/vm.rs
+        
+        #[cfg(target_arch = "x86_64")]
+        MpTable::new(vm.config.num_vcpus)?.write(guest_memory)?;
+        ```
+
+       2. Create KVM `irqchip`. This creates the virtual IOAPIC and virtual
+        PIC and sets up future vCPUs for local APIC.
+
+        ```rust
+        // src/vm-vcpu/src/vm.rs
+      
+        #[cfg(target_arch = "x86_64")]
+        vm.setup_irq_controller()?;
+        ```
+
+       3. Create vCPUs. An `fd` is registered with KVM for each vCPU.
+      
+        ```rust
+        // src/vm-vcpu/src/vm.rs
+        
+        vm.create_vcpus(bus, vcpus_config, guest_memory)?;
+        ```
+
+       4. Configure CPUID. Required (at least) because it’s the means by which
+        the guest finds out it’s virtualized.
+      
+        ```rust
+        // src/vm-vcpu/src/vcpu/mod.rs
+        
+        #[cfg(target_arch = "x86_64")]
+        let base_cpuid = _kvm
+        .get_supported_cpuid(kvm_bindings::KVM_MAX_CPUID_ENTRIES)
+        .map_err(Error::KvmIoctl)?;
+        ```
+
+       5. Configure MSRs (model specific registers). These registers control
+        (among others) the processor features. See the
+        [reference](https://www.intel.co.uk/content/dam/www/public/us/en/documents/manuals/64-ia-32-architectures-software-developer-system-programming-manual-325384.pdf#G14.8720).
+      
+        ```rust
+        // src/vm-vcpu/src/vcpu/mod.rs
+        #[cfg(target_arch = "x86_64")]
+        let supported_msrs = msrs::supported_guest_msrs(_kvm).map_err(Error::GetSupportedMsrs)?;
+        ```
+
+       6. Configure other registers (`kvm_regs`, `kvm_sregs`, `fpu`) and the
+        LAPICs.
+
+        ```rust
+        // src/vm-vcpu/src/vcpu/mod.rs
+      
+        #[cfg(target_arch = "x86_64")]
+        {
+            vcpu.configure_cpuid(&vcpu.config.cpuid)?;
+            vcpu.configure_msrs()?;
+            vcpu.configure_sregs(memory)?;
+            vcpu.configure_lapic()?;
+            vcpu.configure_fpu()?;
+        }
+        ```
+
+
+
+5. Configure legacy devices. This is done partially through `kvm-ioctls`,
```

Comment:
You could take a look at [dbs-legacy-devices](https://github.com/openanolis/dragonball-sandbox/tree/main/crates/dbs-legacy-devices)